pkg/reftracker/ref_tracker.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package reftracker
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+type AppRefTracker struct {
+	lock       sync.Mutex
+	refsToApps map[string]map[string]struct{}
+	appsToRefs map[string]map[string]struct{}
+}
+
+func NewAppRefTracker() *AppRefTracker {
+	return &AppRefTracker{refsToApps: map[string]map[string]struct{}{}, appsToRefs: map[string]map[string]struct{}{}}
+}
+
+func (a AppRefTracker) AddAppForRef(resourceKind, refName, namespace, appName string) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	refKey := fmt.Sprintf(`%s:%s:%s`, resourceKind, refName, namespace)
+	apps := a.refsToApps[refKey]
+	if apps == nil {
+		apps = map[string]struct{}{}
+	}
+
+	appKey := fmt.Sprintf(`%s:%s`, appName, namespace)
+	refs := a.appsToRefs[appKey]
+	if refs == nil {
+		refs = map[string]struct{}{}
+	}
+
+	apps[appName] = struct{}{}
+	a.refsToApps[refKey] = apps
+
+	appsRefKey := fmt.Sprintf(`%s:%s`, refName, resourceKind)
+	refs[appsRefKey] = struct{}{}
+	a.appsToRefs[appKey] = refs
+}
+
+func (a AppRefTracker) AppsForRef(resourceKind, resourceName, namespace string) (map[string]struct{}, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	refKey := fmt.Sprintf(`%s:%s:%s`, resourceKind, resourceName, namespace)
+	if a.refsToApps[refKey] == nil {
+		return nil, fmt.Errorf("could not find App ref %s/%s", resourceKind, resourceName)
+	}
+
+	return a.refsToApps[refKey], nil
+}
+
+func (a AppRefTracker) RemoveRef(resourceKind, resourceName, namespace string) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	refKey := fmt.Sprintf(`%s:%s:%s`, resourceKind, resourceName, namespace)
+	delete(a.refsToApps, refKey)
+}
+
+func (a AppRefTracker) RemoveAppFromAllRefs(refs map[string]struct{}, resourceKind, namespace, appName string) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	for ref := range refs {
+		refKey := fmt.Sprintf(`%s:%s:%s`, resourceKind, ref, namespace)
+		apps := a.refsToApps[refKey]
+		if apps == nil {
+			continue
+		}
+
+		delete(apps, appName)
+		a.refsToApps[refKey] = apps
+	}
+
+	appKey := fmt.Sprintf(`%s:%s`, appName, namespace)
+	delete(a.appsToRefs, appKey)
+}
+
+func (a AppRefTracker) PruneAppFromRefs(currentRefs map[string]struct{}, resourceKind, namespace, appName string) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	appKey := fmt.Sprintf(`%s:%s`, appName, namespace)
+	refsInState := a.appsToRefs[appKey]
+
+	// Compare current state against App's
+	// previous refs.
+	var diff []string
+	for ref := range refsInState {
+		// The format of ref is refName:kind
+		// so we need to get the refName portion.
+		index := strings.Index(ref, ":")
```

Comment:
it seems that in quite a few places we use string concatenation to build a key. typically that's fine but as soon as we start unpacking that string, we should switch to a struct to hold this info. additionally all of the methods on this class take a lot of strings which is pretty error-prone to ordering mistakes. let's have a few tiny structs that represent what we need to key off things. we can also use them as args to functions to make it easier to distinguish fields.

Reply:
Created a `RefKey` struct that is a combo of the resource kind, name, and namespace to address this.